components/i2cdev/i2cdev_master.c

```diff
@@ -0,0 +1,197 @@
+        ESP_LOGD(TAG, "I2C driver successfully reconfigured on port %d", dev->port);
+    }
+    return i2c_setup_device(dev);
+}
+
+esp_err_t _i2cdev_init()
+{
+    memset(states, 0, sizeof(states));
+
+#if !CONFIG_I2CDEV_NOLOCK
+    for (int i = 0; i < I2C_NUM_MAX; i++)
+    {
+        states[i].lock = xSemaphoreCreateMutex();
+        if (!states[i].lock)
+        {
+            ESP_LOGE(TAG, "Could not create port mutex %d", i);
+            return ESP_FAIL;
+        }
+    }
+#endif
+
+    return ESP_OK;
+}
+
+esp_err_t _i2cdev_done()
+{
+    for (int i = 0; i < I2C_NUM_MAX; i++)
+    {
+        if (!states[i].lock) continue;
+
+        if (states[i].installed)
+        {
+            SEMAPHORE_TAKE(i);
+            if (!states[i].bus)
+            {
+                ESP_LOGW(TAG, "The state of I2C port %d is `installed`, but bus handle is NULL)", i);
+            }
+            else
+            {
+                i2c_del_master_bus(states[i].bus);
+            }
+            states[i].installed = false;
+            SEMAPHORE_GIVE(i);
+        }
+#if !CONFIG_I2CDEV_NOLOCK
+        vSemaphoreDelete(states[i].lock);
+#endif
+        states[i].lock = NULL;
+    }
+    return ESP_OK;
+}
+
+esp_err_t _i2c_dev_probe(i2c_dev_t *dev, i2c_dev_type_t operation_type)
+{
+    if (!dev) return ESP_ERR_INVALID_ARG;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    esp_err_t res = i2c_setup_port(dev);
+    if (res == ESP_OK)
+    {
+        res = i2c_master_probe(dev->bus, dev->device_config.device_address, CONFIG_I2CDEV_TIMEOUT);
+    }
+
+    SEMAPHORE_GIVE(dev->port);
+
+    return res;
+}
+
+esp_err_t _i2c_dev_read(i2c_dev_t *dev, const void *out_data, size_t out_size, void *in_data, size_t in_size)
+{
+    if (!dev || !in_data || !in_size) return ESP_ERR_INVALID_ARG;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    esp_err_t res = i2c_setup_port(dev);
+    if (res == ESP_OK)
+    {
+
+        if (out_data && out_size)
+        {
+            res = i2c_master_transmit_receive(dev->device, (void *)out_data, out_size, in_data, in_size, CONFIG_I2CDEV_TIMEOUT);
+        }
+        else
+        {
+            res = i2c_master_receive(dev->device, (void *)out_data, out_size, CONFIG_I2CDEV_TIMEOUT);
+        }
+        if (res != ESP_OK)
+        {
+            ESP_LOGE(TAG, "Could not read from device [0x%02x at %d]: %d (%s)", dev->addr, dev->port, res, esp_err_to_name(res));
+        }
+    }
+
+    SEMAPHORE_GIVE(dev->port);
+    return res;
+}
+
+esp_err_t _i2c_dev_write(i2c_dev_t *dev, const void *out_reg, size_t out_reg_size, const void *out_data, size_t out_size)
+{
+    if (!dev || !out_data || !out_size) return ESP_ERR_INVALID_ARG;
+    uint8_t *write_buffer = NULL;
+    size_t write_buffer_size = 0;
+    esp_err_t res = ESP_FAIL;
+    bool with_register_address = out_reg && out_reg_size;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    write_buffer_size = with_register_address ? out_reg_size + out_size : out_size;
+    write_buffer = malloc(write_buffer_size);
```

Comment:
I am guessing you're using a heap allocated buffer to prevent a task's (limited) stack from being used instead. Is it not possible to use the stack?

Reply:
As the part is obtained from the old code, I cannot comment on this.

Reply:
Hmmm, then it would depend on the stack size of the task, set up by the user. Most of the tasks have it set to `configMINIMAL_STACK_SIZE * 8 or 3`, not sure if that would be enough for a temp buffer.